### PR TITLE
fix(plugins): allow for plugin configs on all services

### PR DESCRIPTION
**Pull request by link108**
Halyard does not currently have knowledge of which plugins should be placed onto each spinnaker service. For now, we can log a warning if there are plugin configs placed onto services that are not required for the plugin to function, so that halyard can successfully deploy plugins to Spinnaker.
Currently Spinnaker throws an error if plugin configs are laid down upon a service that does not require it

### CODE REVIEW

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt [hunk 1]
```diff
@@ -127,4 +131,14 @@ open class SpinnakerPluginManager(
     .add(super.createPluginRepository())
 
   override fun getPluginFactory(): PluginFactory = spinnakerPluginFactory
+
+  // TODO (link108): remove this override, once plugin deployments via halyard are fixed
+  override fun loadPlugin(pluginPath: Path?): String? {
```
Comment (link108) [inline kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt]: add a config / flag to allow for strict mode

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/bundle/PluginBundleExtractor.kt [hunk 2]
```diff
@@ -55,12 +58,20 @@ class PluginBundleExtractor {
...
-    // bundles should actually be downloaded, or B) The plugin author incorrectly identified this [service] as one
-    // that the plugin extends (via the PluginInfo `requires` list).
-    throw IntegrationException("Downloaded plugin bundle does not have plugin for service '$service'")
+    if (isStrictPluginLoading()) {
+      // If thrown, this is an indicator that either: A) There's a bug in the plugin framework resolving which plugin
+      // bundles should actually be downloaded, or B) The plugin author incorrectly identified this [service] as one
+      // that the plugin extends (via the PluginInfo `requires` list).
+      throw IntegrationException("Downloaded plugin bundle does not have plugin for service '$service'")
+    } else {
+      log.warn("Downloaded plugin bundle: {}, does not have plugin for service: {}", bundlePath.fileName, service)
```
Comment (robzienert) [inline kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/bundle/PluginBundleExtractor.kt]:
```suggestion
      log.warn("Downloaded plugin bundle '{}' does not have plugin for service: {}", bundlePath.fileName, service)
```

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt [hunk 3]
```diff
@@ -127,4 +133,14 @@ open class SpinnakerPluginManager(
...
+    require(!(pluginPath == null || Files.notExists(pluginPath))) { String.format("Specified plugin %s does not exist!", pluginPath) }
```
Comment (robzienert) [inline kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt]:
```suggestion
    require(!(pluginPath == null || Files.notExists(pluginPath))) { "Specified plugin '$pluginPath' does not exist!" }
```

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt [hunk 4]
```diff
@@ -127,4 +133,14 @@ open class SpinnakerPluginManager(
...
+    log.debug("Loading plugin from '{}'", pluginPath)
+    val pluginWrapper = loadPluginFromPath(pluginPath) ?: return null
+    // try to resolve  the loaded plugin together with other possible plugins that depend on this plugin
+    resolvePlugins()
+    return pluginWrapper!!.descriptor.pluginId
```
Comment (robzienert) [inline kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt]: `!!` is a code smell, let's avoid it when it's easy to do so:

```kotlin
return loadPluginFromPath(pluginPath)
  ?.let {
    // try to resolve  the loaded plugin together with other possible plugins that depend on this plugin
    resolvePlugins()
    it.descriptor.pluginId
  }
```

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt [hunk 5]
```diff
@@ -54,16 +57,19 @@ open class SpinnakerPluginManager(
...
 
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
   private val springExtensionFactory: ExtensionFactory = SpinnakerExtensionFactory(
     this,
     configFactory,
     sdkFactories
   )
-  private val bundleExtractor = PluginBundleExtractor()
+  private val bundleExtractor = PluginBundleExtractor(environment)
```
Comment (jonsie) [inline kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt]: Nit: Mayne pass an instance of `PluginBundleExtractor` to the `SpinnakerPluginManager` and avoid adding a new member field `environment` which has the sole purpose of being used as a constructor arg for `PluginBundleExtractor`.

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt (hunk 4 above)
Comment (link108) [reply kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt]: makes sense, updating 👍

File: kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt (hunk 5 above)
Comment (link108) [reply kork-plugins/src/main/kotlin/com/netflix/spinnaker/kork/plugins/SpinnakerPluginManager.kt]: totally, I like that approach 👍

**Pull request merged**